src/ADTypes.jl

```diff
@@ -46,9 +47,21 @@ export AutoChainRules,
        AutoTapir,
        AutoTracker,
        AutoZygote
-
 export AutoSparse
 
+# Sparsity detection
+export AbstractSparsityDetector
```

Comment:
don't export this, most users won't need it

Reply:
1. I think the first use-case is quite similar the one of `AbstractADType`: writing an interface that allows users to specify the sparsity detector

2. It's mentioned in the docstrings of the exported interface functions:
  ```markdown
    jacobian_sparsity(f, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
    jacobian_sparsity(f!, y, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
  
  Use detector sd to construct a (typically sparse) matrix S describing the pattern of nonzeroes in the Jacobian of f (resp. f!) applied
  at x (resp. (y, x)).
  ```

3. It makes it easy for users to find sparsity detectors:
  ```Julia-repl
  julia> subtypes(AbstractSparsityDetector)
  3-element Vector{Any}:
   ADTypes.KnownHessianSparsityDetector
   ADTypes.KnownJacobianSparsityDetector
   ADTypes.NoSparsityDetector
  ```
  
4. It makes stack-traces from method errors a bit shorter.


I'm not sure what is being gained by not exporting this.

Reply:
I think exports should be the stuff that people use all the time, and I don't see `AbstractSparsityDetector` being used often. Even for people who wonder which sparsity detectors are available, they'll probably check out the docstring of `jacobian_sparsity` before they remember they can use `subtypes` and other discovery tooling

Reply:
>  Even for people who wonder which sparsity detectors are available, they'll probably check out the docstring of `jacobian_sparsity` before they remember they can use subtypes and other discovery tooling

The docstring of `jacobian_sparsity` is my point 2 above. There are no references of any sparsity detectors in that docstring, only of a (currently unexpected) `AbstractSparsityDetector`. The next logical step for a user is to type `? AbstractSparsityDetector` in the REPL and get an error:

```Julia-repl
julia> using ADTypes

help?> AbstractSparsityDetector
search:

Couldn't find AbstractSparsityDetector
Perhaps you meant AbstractVector
  No documentation found.

  Binding AbstractSparsityDetector does not exist.
```

Reply:
But I don't feel strongly about it so I'll let someone else make the call.

Reply:
fair enough, let's keep it